### PR TITLE
feat(042): approvable tier changes on active licence assignments

### PR DESCRIPTION
**Pull request by studert**
## Summary

A tier-upgrade request arrived for a licence someone already held, and the Hub had no way to action it. Approving it failed outright:

> Requester already has an active assignment for this tool (assignment #N). Revoke the existing assignment first, or cancel this request.

Revoke-then-recreate wasn't a workaround — it orphans the comment thread, resets `assigned_at`, drops the stored API key, and **double-bills the person in the switch month**.

The root problem turned out not to be a missing feature but **three unreconciled implementations** of the same operation. This unifies them and makes "change the tier of an active assignment" approvable. **No schema change, no migration.**

Spec, plan and full running log: [`specs/042-assignment-tier-change/`](specs/042-assignment-tier-change/) — `implementation-plan.html` for the design, `implementation-notes.html` for the challenge pass, deviations, corrections and verification.

### Licence-request approval now has three modes

`approveRequestSchema` becomes a **discriminated union**, not a flat flag:

| mode | Requester already holds the tool | Outcome |
|---|---|---|
| `create` | no | Insert, exactly as before |
| `create` | yes | Refused (advice updated: retier instead of revoke) |
| `change_tier` | yes, manual seat | **New** — retier in place, link the request |
| `change_tier` | yes, same tier | No-op on the row, but **still** approves and links |
| `link_existing` | yes | **New** — approve and link, mutate nothing |

A union rather than a flag because `loadToolAndTier` runs *before* the transaction and unconditionally demanded an API key — so create-only rules had to be **unreachable** from the other modes, not merely skipped. Claude Console is the only `requires_api_key` tool *and* has the deepest tier ladder, i.e. exactly the upgrades a shared shape would have broken.

`link_existing` is what gives **GitHub Copilot** an approvable outcome at all: all 63 live seats are sync-managed, so a Business→Enterprise request was refused by the create guard *and* by the sync guard. Now it matches 032-v2's provision-first philosophy — upgrade the plan in GitHub, approve as linked, the tier arrives via sync.

### Why the tier change mutates in place

Per-period expected spend sums `cost_at_assignment_cents` by date overlap, flat monthly price, **no proration**. Closing one row and opening another therefore double-counts the switch period — one seat, two prices.

Both options are wrong, in different places:

| | Error | Where |
|---|---|---|
| Close old + open new | `1 period × old price` | The switch month only |
| **In place** (chosen) | `N closed periods × delta` | Every closed period the row spans |

In-place is the **larger** total error. It was chosen anyway because it is the convention this codebase already committed to — `actions/tools.ts` states it in a comment and migration `0024` backfilled production to match — and because the double count corrupts the *current* month, the figure people act on. A feature whose purpose is to reconcile three rival semantics shouldn't introduce a fourth costing model. Confirmed with the requester after the trade-off was laid out; alternatives (boundary-snapped supersede, proration, freezing expected spend at period close) are on record as deferral D-1.

### Bugs fixed along the way

- **The assignment detail page crashed for every sync-managed seat** — `Tooltip` used without a `TooltipProvider`; the assignments *table* only works because `DataTable` happens to supply one. Found by the browser pass.
- **The tier control rendered empty on every assignment** — radix resolves a `Select`'s display value from the selected *item's* children, which load asynchronously. Pre-existing, but this feature's primary control.
- **The detail page's tier `<Select>` was unguarded against Copilot sync** — an admin could change a synced seat's tier, see `[SAVED]`, and have cron silently revert it. The gate is **tool-based, not `source`-based**, because sync takes over manual rows too.
- **`loadToolAndTier` never checked `access_tiers.is_active`** — the request path could put a live seat on a deactivated tier, which both other entry points already refuse.
- **The tier write had no `status` guard** — a concurrent revoke could be retiered, corrupting the historical snapshot that revoked rows are supposed to preserve. Now a conditional UPDATE on observed `status` + `tier_id` (which doubles as an optimistic lock against the price cascade) with a zero-rows sentinel.
- **Cost invalidation was both wrong and wasteful** — it duplicated `updateTier`'s path list (already drifted) and fired on every edit, including workspace/API-key-only changes and no-ops. Now one shared list, gated on cost actually changing.
- **Dead code removed** — `getExpectedSpendForPeriod` had zero callers and carried a `status` filter the surviving readers don't, inviting a future "reconcile these two" against a function nobody read.

### Notable non-changes

Tier semantics live in one pure module (`lib/assignments/tier-change.ts`) shared by `updateAssignment` and `approveRequest`. No new action, no new dialog, and no `tier_changed_at` column — `change_history` already stores old/new `tierId` with actor and timestamp, and now backs a tier timeline on the assignment page.

## Verification

| Gate | Result |
|---|---|
| `pnpm typecheck` | clean |
| `pnpm test` | **659 passed**, 53 files (from 629) |
| `pnpm lint` / eslint on changed files | clean |
| `pnpm test:integration` (full directory) | **8 files, 56 passed**, 7 todo, 2 skipped |
| Playwright browser pass | **9/9 checks** |

Integration and browser runs were against a **branch of the real database** — 223 users, 5 tools, 309 active assignments, 27 requests, 1933 change-history rows.

**The double-count property is pinned by a *unit* test.** The overlap predicate was extracted from `getBudgetWithCosts` specifically so it runs in CI: the `integration-tests` job in `.github/workflows/ci.yml` is commented out, so an integration test could never have gated a merge. `fetchPerToolByPeriod` now shares that predicate too, with its long-standing `>` vs `>=` edge case preserved as an explicit parameter and its own test.

**Browser pass, end to end:** a real Claude seat moved Standard Seat → Premium Seat → Standard Seat (both directions); header and cost followed ($25.00 ↔ $125.00); feedback as inline `[SAVED]`, not a toast; the timeline rendered both hops attributed and dated. Confirmed directly in the database afterwards: **exactly one active row** for that user+tool, with `tierId` 5→6 and `costAtAssignmentCents` 2500→12500 written under a single timestamp. A real `copilot-sync` seat showed "Managed by sync", tier control disabled but still legible, and no re-pricing disclosure.

**Two integration-test defects the run exposed**, both fixed here: this spec's audit rows broke the suite's teardown via `change_history.changed_by`'s `onDelete RESTRICT` (the suite reported *12/12 tests passing and a failed file*); and one of my own assertions tested the wrong layer, expecting `approveRequest` to render a template it deliberately stores verbatim.

**Process:** plan → 5-lens adversarial challenge pass (2 blocking + several major findings adopted; scope roughly halved, and it disproved the plan's original costing argument) → implement → `/simplify` (4 lenses, 10 applied, 4 skipped with reasons) → verify. All logged in `implementation-notes.html`, including two claims I had to retract.

**Known limitations:** the approval dialog's three modes are covered server-side but weren't clicked through in a browser. Copilot's sync-managed path was verified as *gated*, not as an end-to-end GitHub round trip.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/actions/license-requests.ts [hunk 1]
```diff
@@ -335,20 +491,91 @@ export async function approveRequest(
...
+        if (approval.mode === "change_tier") {
+          const outcome = buildTierChange(
+            {
+              tierId: target.tierId,
+              costAtAssignmentCents: target.costAtAssignmentCents,
+              // Already refused above; the tool cannot change mid-transaction.
+              isSyncManaged: false,
+            },
+            { id: tier!.id, monthlyCostCents: tier!.monthlyCostCents },
+          );
```
[the remark this answers is not in the document]
Comment (studert) [reply src/actions/license-requests.ts:568]: Fixed in `f4bb80b` — you're right, and this was a real defect, not just a hardening gap.

`loadToolAndTier` proves the *tier* belongs to `approval.toolId`, but nothing tied the *assignment* to it, so the two were free to disagree. Both consequences you name are real: a stale dialog or crafted payload could write another tool's tier onto the seat, and because the sync-managed refusal also keys off the selected tool rather than the row's own, passing a Copilot `assignmentId` with a non-Copilot `toolId` would have walked straight past that gate.

Added the explicit `target.toolId === approval.toolId` check before `buildTierChange`, which mirrors the same-tool rule `updateAssignment` already gets for free by scoping its `accessTiers` lookup to `assignment.toolId` — splitting that validation across two loads is exactly what opened the hole. Also added a unit regression test that asserts both the refusal and that nothing is written to the assignment.

Worth noting the adversarial challenge pass missed this: it scrutinised the mode matrix and the sync gate separately, but never their interaction. Recorded in `implementation-notes.html`.

File: src/actions/history.ts [hunk 2]
```diff
@@ -115,3 +115,64 @@ export async function getEntityHistory(
...
+): Promise<ActionResult<TierChangeEntry[]>> {
+  const records = await db.query.changeHistory.findMany({
+    where: and(
+      eq(changeHistory.entityType, "license_assignment"),
+      eq(changeHistory.entityId, assignmentId),
+      eq(changeHistory.fieldName, "tierId"),
+    ),
+    orderBy: desc(changeHistory.createdAt),
+    with: { changedByUser: true },
+  });
```
[the remark this answers is not in the document]
Comment (studert) [reply src/actions/history.ts:148]: Fixed in `f4bb80b` — restricted to `with: { changedByUser: { columns: { name: true } } }`, since `name` is the only field rendered.

One clarification on severity, so the record is accurate: the mapped return shape is `{ id, previousTierName, newTierName, changedByName, createdAt }`, so `passwordHash` never reached the client. It was needless loading of a password hash into server memory rather than an exposure. Still worth fixing, and it now matches the `columns: {...}` convention used elsewhere in this file.

File: src/actions/assignments.ts [hunk 3]
```diff
@@ -294,22 +316,57 @@ export async function updateAssignment(
...
-    .set(updateValues)
-    .where(eq(licenseAssignments.id, id));
+  // 042: guard the write on the state we actually read. Five concurrent
+  // producers can flip this row to inactive between the read above and here
+  // (revokeLicense, assignLicense's deactivate leg, deactivateUser's bulk
+  // revoke, copilot-sync's removed-seat revoke), and retiering a just-revoked
+  // row would corrupt the historical snapshot that revoked rows are supposed to
+  // preserve. The tierId predicate doubles as an optimistic lock against a
+  // concurrent tier price cascade (updateTier / syncBillingData) leaving the row
+  // stranded at a stale price.
```
[the remark this answers is not in the document]
Comment (studert) [reply src/actions/assignments.ts]: Fixed in `f4bb80b` — you're right, the comment was wrong and I've rewritten it.

Confirmed against `updateTier` (`src/actions/tools.ts`): the cascade sets `costAtAssignmentCents` on rows matched by `tierId`, leaving `tier_id` itself untouched — so an `eq(tierId, observed)` predicate cannot see it. Same for `syncBillingData`.

The comment now states only what the guard actually covers (a status flip to inactive from any of the four revoke paths, and a concurrent retier), then names the price-cascade race it does **not** cover and why it is tolerable: our cost value is read before the transaction and would win, but the error is narrow and self-correcting — the next cascade or retier overwrites it. A real fix is row-level compare-and-swap on `updatedAt`, which the plan deliberately scoped out (the guard was scoped to the tier-snapshot invariant); that trade-off is recorded in `implementation-notes.html` under the simplify pass's skipped findings.
